### PR TITLE
Add obsolete attribute to DisablePrivateReflectionAttribute

### LINKED ISSUE #43247: API proposal: obsolete DisablePrivateReflectionAttribute
Per https://github.com/dotnet/runtime/issues/11811, the runtime does not honor `DisablePrivateReflectionAttribute`, and there are no plans to restore this capability. We should obsolete the API.

```cs
namespace System.Runtime.CompilerServices
{
    [Obsolete("DisablePrivateReflectionAttribute has no effect in .NET Core and .NET 5.0+ applications.")] // NEW attribute
    [AttributeUsage(AttributeTargets.Assembly, AllowMultiple = false, Inherited = false)]
    public sealed class DisablePrivateReflectionAttribute : Attribute
    {
        public DisablePrivateReflectionAttribute() { }
    }
}
```

I recommend using a normal `[Obsolete]` attribute with no diagnostic id since there's no real required reading here, nor should people really be suppressing this. They should just remove the attribute. There is no replacement mechanism to restore this functionality.

If we really wanted a diagnostic id and the whole breaking change rigamarole that it entails, we could use `SYSLIB0013`. I'll file a separate doc issue regardless of whether we use a unique diagnostic id here.

### PR DESCRIPTION
**Pull request by annchous**
Close #43247

### CODE REVIEW

**Comment by annchous** [comment]
It seems to me that a more appropriate tag is `area-System.Runtime.CompilerServices`
cc: @stephentoub

**Comment by annchous** [comment]
cc: @stephentoub @jkotas

File: src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/DisablePrivateReflectionAttribute.cs [hunk 1]
```diff
@@ -3,6 +3,7 @@
 
 namespace System.Runtime.CompilerServices
 {
+    [Obsolete("DisablePrivateReflectionAttribute has no effect in .NET Core and .NET 5.0+ applications.")]
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/DisablePrivateReflectionAttribute.cs:6]: Per https://github.com/dotnet/runtime/issues/43247#issuecomment-713070795, this needs an ID set on it. That ID will need to be recorded in https://github.com/dotnet/runtime/blob/0c3cc15f90f0e3b81ce3199be1ac217564043635/src/libraries/Common/src/System/Obsoletions.cs.  You can see other examples in that file and how they're used in call sites.

In addition, this is editing the src file, but the class in the reference assembly file https://github.com/dotnet/runtime/blob/79ae74f5ca5c8a6fe3a48935e85bd7374959c570/src/libraries/System.Runtime/ref/System.Runtime.cs will also need to have the attribute applied.

File: src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/DisablePrivateReflectionAttribute.cs [hunk 2]
```diff
@@ -3,6 +3,7 @@
 
 namespace System.Runtime.CompilerServices
 {
+    [Obsolete(Obsoletions.EscapeUriStringMessage, DiagnosticId = Obsoletions.EscapeUriStringDiagId, UrlFormat = Obsoletions.SharedUrlFormat)]
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/DisablePrivateReflectionAttribute.cs:6]: This needs a new message and new diagnostic ID for this particular obsoletion added to that file, and then those new values used here.  It shouldn't use the existing ones, which aren't relevant to this API.

File: src/libraries/System.Runtime/ref/System.Runtime.cs [hunk 3]
```diff
@@ -9381,6 +9381,7 @@ public sealed partial class DependencyAttribute : System.Attribute
         public string DependentAssembly { get { throw null; } }
         public System.Runtime.CompilerServices.LoadHint LoadHint { get { throw null; } }
     }
+    [System.ObsoleteAttribute("DisablePrivateReflectionAttribute has no effect in .NET Core and .NET 5.0+ applications.", DiagnosticId = "SYSLIB0013", UrlFormat = "https://aka.ms/dotnet-warnings/{0}")]
```
Comment (stephentoub) [inline src/libraries/System.Runtime/ref/System.Runtime.cs:9384]: SYSLIB0013 is already taken.  A new obsoletion ID is needed.

Comment (annchous) [reply src/libraries/System.Runtime/ref/System.Runtime.cs:9384]: It was mentioned in the commentary on this issue, so I decided that it was necessary to use it. Thanks for the explanation. Can I take SYSLIB0015 instead?

Comment (annchous) [reply src/libraries/System.Runtime/ref/System.Runtime.cs]: I hope I did it right :)

**Comment by annchous** [comment]
@stephentoub anything else to add or correct?

**Comment by annchous** [comment]
@stephentoub should I contribute the new obsoletion ID here https://github.com/dotnet/runtime/blob/main/docs/project/list-of-obsoletions.md?

**Comment by stephentoub** [comment]
> should I contribute the new obsoletion ID here

Yes, please. Thanks.

**Review by stephentoub — APPROVED** [review]
Thanks.

**Comment by annchous** [comment]
@stephentoub I've finished with docs. Thank you too.

**Comment by annchous** [comment]
I see that some of the checks failed (exactly half). It is necessary to somehow mark in which places there should be no warnings that the attribute is now obsolete. How should this be done?

File: src/libraries/Common/src/System/Obsoletions.cs [hunk 4]
```diff
@@ -48,5 +48,8 @@ internal static class Obsoletions
 
         internal const string WebRequestMessage = "Use HttpClient instead.";
         internal const string WebRequestDiagId = "SYSLIB0014";
+
+        internal const string DisablePrivateReflectionAttributeMessage = "DisablePrivateReflectionAttribute has no effect in .NET Core and .NET 5.0+ applications.";
```
Comment (jkotas) [inline src/libraries/Common/src/System/Obsoletions.cs:52]: This attribute still has effect in a few corner cases in CoreCLR. Search for `DisablePrivateReflection` under `src\coreclr` and delete everything found.

Comment (jkotas) [reply src/libraries/Common/src/System/Obsoletions.cs:52]: Also, we may want to say `.NET 6.0+` in the error message.

**Comment by annchous** [comment]
@jkotas I've finished with changes that you requested. But there is some trouble with building libs:
```
C:\runtime\artifacts\obj\netstandard\net6.0-Debug\netstandard.Forwards.cs(1524,67): error SYSLIB0015: "DisablePrivateReflectionAttribute" is obsolete: 
'DisablePrivateReflectionAttribute has no effect in .NET Core and .NET 5.0+ applications.' [C:\runtime\src\libraries\shims\generated\netstandard.csproj]
```
How can this be fixed?

**Comment by jkotas** [comment]
Could you please also delete this: https://github.com/dotnet/runtime/blob/main/src/coreclr/inc/corhdr.h#L1824-L1827

**Comment by jkotas** [comment]
> How can this be fixed?

Try to add `<NoWarn>$(NoWarn);SYSLIB0003</NoWarn>` to src\libraries\shims\generated\netstandard.csproj .

**Comment by annchous** [comment]
> Try to add `<NoWarn>$(NoWarn);SYSLIB0015</NoWarn>` to src\libraries\shims\generated\netstandard.csproj .

Yes, I've tried that. This cleared the error from netstandard, but followed in turn by the same error from mscorelib (`src\libraries\shims\generated\mscorlib.csproj`). If you insert NoWarn in `src\libraries\shims\generated\mscorlib.csproj`, then the error occurs in `System.Runtime.csproj`.

**Comment by jkotas** [comment]
Ok, then try this central place instead: https://github.com/dotnet/runtime/blob/main/src/libraries/Directory.Build.targets#L20

**Comment by annchous** [comment]
@jkotas I've tried to add `<NoWarn>$(NoWarn);SYSLIB0015</NoWarn>` to every project file I mentioned above. Now it builds successfully, but is it ok to do like that?

**Comment by jkotas** [comment]
I think it is ok.

**Comment by annchous** [comment]
> Ok, then try this central place instead: https://github.com/dotnet/runtime/blob/main/src/libraries/Directory.Build.targets#L20

@jkotas That works nice and looks better than adding NoWarn to every project file. I will commit this change. Thank you!

**Comment by annchous** [comment]
@jkotas I think I have definitely finished working on this now. The checks are going well. Anything else?

File: src/libraries/Common/src/System/Obsoletions.cs [hunk 5]
```diff
@@ -48,5 +48,8 @@ internal static class Obsoletions
 
         internal const string WebRequestMessage = "Use HttpClient instead.";
         internal const string WebRequestDiagId = "SYSLIB0014";
+
+        internal const string DisablePrivateReflectionAttributeMessage = "DisablePrivateReflectionAttribute has no effect in .NET Core and .NET 6.0+ applications.";
```
Comment (stephentoub) [inline src/libraries/Common/src/System/Obsoletions.cs:52]:
```suggestion
        internal const string DisablePrivateReflectionAttributeMessage = "DisablePrivateReflectionAttribute has no effect in .NET 6.0+ applications.";
```

File: src/libraries/System.Runtime/ref/System.Runtime.cs [hunk 6]
```diff
@@ -9381,6 +9381,7 @@ public sealed partial class DependencyAttribute : System.Attribute
         public string DependentAssembly { get { throw null; } }
         public System.Runtime.CompilerServices.LoadHint LoadHint { get { throw null; } }
     }
+    [System.ObsoleteAttribute("DisablePrivateReflectionAttribute has no effect in .NET Core and .NET 6.0+ applications.", DiagnosticId = "SYSLIB0015", UrlFormat = "https://aka.ms/dotnet-warnings/{0}")]
```
Comment (stephentoub) [inline src/libraries/System.Runtime/ref/System.Runtime.cs:9384]:
```suggestion
    [System.ObsoleteAttribute("DisablePrivateReflectionAttribute has no effect in .NET 6.0+ applications.", DiagnosticId = "SYSLIB0015", UrlFormat = "https://aka.ms/dotnet-warnings/{0}")]
```

**Review by jkotas — APPROVED** [review]
Thank you!

**Review by stephentoub — APPROVED** [review]

**Pull request merged**